### PR TITLE
chore(ssa): Add intial control flow graph 

### LINKED ISSUE #1117: Add control flow graph (CFG)
### Problem

A CFG is a data structure that shows the flow of control in a program.  The nodes of a CFG can be instructions, however since the control flow analysis of instructions that do not affect control flow is uninteresting, we group instructions into basic blocks.

The nodes of the CFG will be basic blocks. The CFG is useful for certain optimizations like loop analysis.

### Proposed solution

Implement CFG data structure in a modular way with tests and comments.

### Alternatives considered

_No response_

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by joss-aztec**
# Related issue(s)

Resolves #1117

# Description

## Summary of changes

Added `ControlFlowGraph` for computing the cfg for a given function.

## Test additions / changes

Unit tests for checking correct cfg computation.

# Checklist

- [x] I have tested the changes locally.
- [ x I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [ ] I have ensured all changes are covered in the description.

## Documentation needs
- [ ] This PR requires documentation updates when merged.

# Additional context

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa_refactor/cfg.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        let successor_node = self.data.entry(to).or_default();
+        assert!(
+            successor_node.predecessors.len() < 2,
+            "ICE: A cfg node cannot have more than two predecessors"
+        );
+        successor_node.predecessors.insert(from);
+    }
+
+    /// Get an iterator over the CFG predecessors to `block`.
+    pub(crate) fn pred_iter(&self, basic_block_id: BasicBlockId) -> PredIter {
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/cfg.rs:88]: Instead of PredIter & SuccIter aliases, can we use `impl Iterator<Item = BasicBlockId>` (or ExactSizeIter) for both?

File: crates/noirc_evaluator/src/ssa_refactor/ir/function.rs [hunk 2]
```diff
@@ -20,6 +23,34 @@ pub(crate) struct Function {
...
+/// An iterator over a function's basic block ids. The iterator type is `BasicBlockId`.
+pub(crate) struct BasicBlockIdsIter<'a>(hash_map::Keys<'a, BasicBlockId, BasicBlock>);
+
+impl<'a> Iterator for BasicBlockIdsIter<'a> {
+    type Item = BasicBlockId;
+
+    fn next(&mut self) -> Option<BasicBlockId> {
+        self.0.next().map(|k| *k)
+    }
+}
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/ir/function.rs:52]: We can avoid adding a new iterator here if we change the function's `basic_blocks` to be a `DenseMap<BasicBlock>`, and add an iter function to that instead.

The iterator there should probably iterate over the id and value, so we can still have the helper function here for mapping the iterator to only the ids.

File: crates/noirc_evaluator/src/ssa_refactor/ir/instruction_predicates.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+/// Visit all successors of a block with a given visitor closure. The closure
+/// arguments are the branch instruction that is used to reach the successor,
+/// and the id of the successor block itself.
+pub(crate) fn visit_block_succs<F: FnMut(BasicBlockId)>(
+    func: &Function,
+    basic_block_id: BasicBlockId,
+    mut visit: F,
+) {
+    let terminator = &func[basic_block_id].terminator();
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/ir/instruction_predicates.rs:13]: terminator already returns a reference so the extra one here is unnecessary

File: crates/noirc_evaluator/src/ssa_refactor/ir/instruction_predicates.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+use super::{
+    super::basic_block::BasicBlockId, function::Function, instruction::TerminatorInstruction,
+};
+
+/// Visit all successors of a block with a given visitor closure. The closure
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/ir/instruction_predicates.rs:5]: The naming of this file is a bit odd to me. Is it intended to group together visitor functions? If so, perhaps we should rename it "visitors" or something more direct. "*_predicates" sounds like a bunch of boolean functions to me.

**Review by jfecher — COMMENTED** [review]
Looks good so far. Let me know when this is ready for review

File: crates/noirc_evaluator/src/ssa_refactor/ir/cfg.rs [hunk 5]
```diff
@@ -86,27 +86,149 @@ impl ControlFlowGraph {
...
+
+    #[test]
+    fn jumps() {
+        let mut func = Function::new();
+        let block0_id = func.entry_block();
+        let cond = func.dfg.add_block_parameter(block0_id, Type::unsigned(1));
+        let block1_id = func.dfg.new_block();
+        let block2_id = func.dfg.new_block();
+
+        func.dfg[block0_id].set_terminator(TerminatorInstruction::JmpIf {
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/ir/cfg.rs:139]: A comment here showing the structure of the ssa you're building could be helpful:
```
// fn func {
// block0(cond: u1):
//     jmpif cond(), then: block2, else: block1
// block1():
//     jmpif cond(), then: block1, else: block2
// block2():
//     return
// }
```

File: crates/noirc_evaluator/src/ssa_refactor/ir/cfg.rs [hunk 6]
```diff
@@ -86,27 +86,149 @@ impl ControlFlowGraph {
...
+
+            assert_eq!(block0_successors.contains(&block1_id), true);
+            assert_eq!(block0_successors.contains(&block2_id), true);
+            assert_eq!(block1_successors.contains(&block1_id), true);
+            assert_eq!(block1_successors.contains(&block2_id), true);
+            assert_eq!(block2_successors, []);
+        }
+
+        // Add a new return block replacing the return in block2
+        let ret_block_id = func.dfg.new_block();
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/ir/cfg.rs:185]:
```
// fn func {
// block0(cond: u1):
//     jmpif cond(), then: block1, else: ret_block
// block1():
//     jmpif cond(), then: block1, else: block2
// block2():
//     jmp ret_block
// ret_block():
//     return
// }
```

**Review by jfecher — APPROVED** [review]
Looks good. We can add the test comments in another PR if needed

File: crates/noirc_evaluator/src/ssa_refactor/ir/cfg.rs (hunk 6 above)
Comment (joss-aztec) [reply crates/noirc_evaluator/src/ssa_refactor/ir/cfg.rs:203]: Thanks - added

File: crates/noirc_evaluator/src/ssa_refactor/ir/dfg.rs [hunk 7]
```diff
@@ -249,15 +247,7 @@ impl std::ops::Index<BasicBlockId> for DataFlowGraph {
     }
 }
 
-impl Index<BasicBlockId> for DataFlowGraph {
-    type Output = BasicBlock;
-    /// Get a reference to a function's basic block for the given id.
-    fn index(&self, id: BasicBlockId) -> &BasicBlock {
-        &self.blocks[id]
-    }
-}
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/ir/dfg.rs]: Why was this removed?

**Review by jfecher — APPROVED** [review]
